semantic/constraints.go

```diff
@@ -306,14 +306,33 @@ func (v ConstraintGenerator) typeof(n Node) (PolyType, error) {
 		if !ok {
 			return nil, errors.New("member object must be a type variable")
 		}
-		v.cs.AddKindConst(tv, KRecord{
+		v.cs.AddKindConst(tv, ObjectKind{
 			properties: map[string]PolyType{n.Property: ptv},
 			lower:      LabelSet{n.Property},
 			upper:      AllLabels(),
 		})
 		return ptv, nil
+	case *IndexExpression:
+		ptv := v.cs.f.Fresh()
+		t, err := v.lookup(n.Object)
+		if err != nil {
+			return nil, err
+		}
+		tv, ok := t.(Tvar)
+		if !ok {
+			return nil, errors.New("object must be a type variable")
+		}
+		idx, err := v.lookup(n.Index)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
I expected an AddTypeConst call here with `idx` and Int. Does this happen inside of the ArrayKind or something? If it does can we move it out? I would expect the ArrayKind to only deal with the element type.

Reply:
The reason I coupled the type constraint and kind constraint together was just in case changes where made to the `ArrayKind` constraint later on. But I agree, the type constraint should be separate.